cmd/promtool/backfill.go

```diff
@@ -0,0 +1,163 @@
+	return &Parser{s: bufio.NewScanner(r)}
+}
+
+// Next advances the parser to the next sample. It returns io.EOF if no
+// more samples were read.
+func (p *Parser) Next() (textparse.Entry, error) {
+	for p.s.Scan() {
+		line := p.s.Bytes()
+		line = append(line, '\n')
+		p.Parser = textparse.New(line, contentTypeLatest)
+		if et, err := p.Parser.Next(); err != io.EOF {
+			return et, err
+		}
+	}
+	if err := p.s.Err(); err != nil {
+		return 0, err
+	}
+	return 0, io.EOF
+}
+
+func getMinAndMaxTimestamps(p textparse.Parser) (int64, int64, error) {
+	var maxt, mint int64 = math.MinInt64, math.MaxInt64
+	for {
+		entry, err := p.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return maxt, mint, errors.Wrap(err, "parse")
+		}
+		if entry != textparse.EntrySeries {
+			continue
+		}
+		l := labels.Labels{}
+		p.Metric(&l)
+		_, ts, _ := p.Series()
+		if *ts > maxt {
+			maxt = *ts
+		}
+		if *ts < mint {
+			mint = *ts
+		}
+		if ts == nil {
+			return maxt, mint, errors.Errorf("Expected timestamp for series %v, got none.", l.String())
+		}
+	}
+	if maxt == math.MinInt64 || maxt <= 0 {
+		maxt = 0
+	}
+	if mint == math.MaxInt64 || mint <= 0 {
+		mint = 0
+	}
+	return maxt, mint, nil
+}
+func createBlocks(input *os.File, mint, maxt int64, outputDir string) error {
+	var offset int64 = 2 * time.Hour.Milliseconds()
+	mint = offset * (mint / offset)
+	blocklen := maxt - mint
+	if blocklen == 0 {
+		return nil
+	}
+	for t := mint; t <= maxt; t = t + offset {
+		w, errw := tsdb.NewBlockWriter(log.NewNopLogger(), outputDir, offset)
+		if errw != nil {
+			return errors.Wrap(errw, "block writer")
+		}
+		ctx := context.Background()
+		app := w.Appender(ctx)
+		_, errReset := input.Seek(0, 0)
+		if errReset != nil {
+			return errors.Wrap(errReset, "seek file")
+		}
+		p2 := NewParser(input)
+		tsUpper := t + offset
+		for {
+			e, err := p2.Next()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return errors.Wrap(err, "parse")
+			}
+			if e != textparse.EntrySeries {
+				continue
+			}
+			l := labels.Labels{}
+			p2.Metric(&l)
+			_, ts, v := p2.Series()
+			if ts == nil {
+				return errors.Errorf("Expected timestamp for series %v, got none.", l.String())
+			}
+			if *ts >= t && *ts < tsUpper {
+				if _, err := app.Add(l, *ts, v); err != nil {
+					return errors.Wrap(err, "add sample")
+				}
+			}
+		}
+		if err := app.Commit(); err != nil {
+			return errors.Wrap(err, "commit")
+		}
+		_, errF := w.Flush(ctx)
+		if errF != nil {
+			return errors.Wrap(errF, "flush")
+		}
+		if errC := w.Close(); errC != nil {
+			return errors.Wrap(errC, "close")
+		}
+	}
+	return nil
+}
+
+func backfill(input *os.File, outputDir string) (err error) {
+	p := NewParser(input)
+	maxt, mint, errTs := getMinAndMaxTimestamps(p)
+	if errTs != nil {
+		return errors.Wrap(errTs, "Error getting min and max timestamp.")
+	}
+	if errCb := createBlocks(input, mint, maxt, outputDir); errCb != nil {
+		os.RemoveAll(outputDir)
```

Comment:
Thinking about it, this is actually quite problematic as users might backfill directly into a TSDB directory that already has blocks (from other backfill command runs or from a Prometheus server acting on the same directory).

I'd say you should study how the TSDB block writer deals with errors (it probably already cleans up remnants). Not deleting blocks that are already correctly written might actually be a desired feature (for long backfills, you don't want to throw away work already done).

Reply:
yep, it does.(https://github.com/prometheus/prometheus/blob/master/tsdb/blockwriter.go#L117) 
Have removed the os.removeall and defered the close() as suggested.